packages/odyssey-react-mui/src/labs/DataComponents/DataView.tsx

```diff
@@ -0,0 +1,387 @@
+  display: "flex",
+  justifyContent: "space-between",
+}));
+
+const AdditionalActionsContainer = styled("div")(() => ({
+  display: "flex",
+  justifyContent: "flex-end",
+}));
+
+const DataView = ({
+  availableLayouts = allAvailableLayouts,
+  bulkActionMenuItems,
+  currentPage = 1,
+  emptyPlaceholder,
+  errorMessage: errorMessageProp,
+  getData: getDataFn,
+  getRowId: getRowIdProp,
+  hasFilters,
+  hasPagination,
+  hasSearch,
+  hasSearchSubmitButton,
+  hasRowReordering,
+  hasRowSelection,
+  initialLayout,
+  isEmpty: isEmptyProp,
+  isLoading: isLoadingProp,
+  isNoResults: isNoResultsProp,
+  isRowReorderingDisabled,
+  noResultsPlaceholder,
+  onChangeRowSelection,
+  onReorderRows,
+  paginationType = "paged",
+  resultsPerPage = 20,
+  searchDelayTime,
+  stackOptions,
+  tableOptions,
+  totalRows,
+}: DataViewProps) => {
+  const odysseyDesignTokens = useOdysseyDesignTokens();
+
+  const [currentLayout, setCurrentLayout] = useState<Layout>(
+    initialLayout ?? availableLayouts[0],
+  );
+
+  const [data, setData] = useState<MRT_RowData[]>([]);
+  const [isLoading, setIsLoading] = useState<boolean>(isLoadingProp ?? true);
+  const [isEmpty, setIsEmpty] = useState<boolean>(isEmptyProp ?? true);
+  const [isNoResults, setIsNoResults] = useState<boolean>(
+    isNoResultsProp ?? false,
+  );
+  const [errorMessage, setErrorMessage] =
+    useState<UniversalProps["errorMessage"]>(errorMessageProp);
+
+  const [search, setSearch] = useState<string>("");
+  const [initialFilters, setInitialFilters] = useState<DataFilter[]>();
+  const [filters, setFilters] = useState<DataFilter[]>();
+
+  const [draggingRow, setDraggingRow] = useState<MRT_Row<MRT_RowData> | null>();
+
+  const [rowSelection, setRowSelection] = useState<MRT_RowSelectionState>({});
+
+  useEffect(() => {
+    onChangeRowSelection?.(rowSelection);
+  }, [rowSelection, onChangeRowSelection]);
+
+  const [pagination, setPagination] = useState({
+    pageIndex: currentPage,
+    pageSize: resultsPerPage,
+  });
+
+  const [tableState, setTableState] = useState<TableState>({
+    columnSorting: [],
+    columnVisibility: {},
+    rowDensity: tableOptions?.initialDensity ?? densityValues[0],
+  });
+
+  const shouldShowFilters = useMemo(
+    () => hasSearch || hasFilters,
+    [hasSearch, hasFilters],
+  );
+
+  const availableFilters = useFilterConversion({
+    filters: filters,
+    columns: tableOptions?.columns,
+  });
+
+  const dataQueryParams = useMemo(
+    () => ({
+      page: pagination.pageIndex,
+      resultsPerPage: pagination.pageSize,
+      search,
+      filters,
+      sort: tableState?.columnSorting,
+    }),
+    [pagination, search, filters, tableState?.columnSorting],
+  );
+
+  const getRowId = getRowIdProp ? getRowIdProp : (row: MRT_RowData) => row.id;
+
+  // Set initial filters
+  useEffect(() => {
+    if (!initialFilters && filters) {
+      setInitialFilters(filters);
+    }
+  }, [filters, initialFilters]);
```

Comment:
If we're only determining the initial filters, we can do this in a `useEffect`, but we could also put it in the `useState` directly. Yes, it would be computed each time, but so is the check on this `useEffect` 🤷.

I have no opinion either way, just making sure you know.

In the case of this one, if we only ever want `initialFilters` to be set once, then I'd create a `useMemo` at the top and set it before the `useState`.

If we do care about this changing, why is it called `initialFilters` and not `filters`? I'm confused.

Reply:
This is here for an annoying reason... `filters` might be set explicitly as a prop, or it might be inferred from the table's `columns`. We can't set `initialFilters` as a memo at the top of the component because we don't necessarily have the available filters yet... we need to run `useFilterConversion` on all the incoming filters and columns first, to mutate them into the proper `DataFilters[]` format.

I tried a bunch of different versions, and did some pair-programming with AI, and the upshot was that it kept coming back to this solution as the best one. (That could mean that I need to rethink how we handle filters entirely, to be fair).